### PR TITLE
Migrate service

### PR DESCRIPTION
**Pull request by masha1997**
Note: Due to KPI threshold resource requires changes (I'll file a bug for kpi_threshold_template, and will implement sharing part in the scope of the fix), so I didn't implement custom_thresholds inside of the service. Worth to mention, that we didn't use it anywhere.

Testing Status:
Status of scrapping, plan is **clean**:
``` terraform
Plan: 4344 to import, 0 to add, 0 to change, 0 to destroy.
```
``` terraform
╷
│ Warning: [ITSI Monitoring - Notable Events and Episodes] Skipping Newly Created Episodes KPI
│ 
│ adhoc KPIs is not supported
╵
/*Similar warnings are skipped, we have only expected ones here*/
╷
│ Warning: [ITSI Monitoring - Framework Health] Skipping Smallest KPI Runtime Headroom % KPI
│ 
│ adhoc KPIs is not supported
╵
╷
│ Warning: Config generation is experimental
│ 
│ Generating configuration during import is currently experimental, and
│ the generated configuration format may change in future versions.
```
Worth to mention, that I received several errors on entity rules. But It is just empty rule, so I believe It is fine:
```terraform
╷
│ Error: Missing Configuration for Required Attribute
│ 
│   with itsi_service.CDN_Outage,
│   on generated.tf line 1:
│   (source code not available)
│ 
│ Must set a configuration value for the
│ entity_rules[Value({"rule":[{"field":<null>,"field_type":"alias","rule_type":"matches","value":<null>}]})].rule[Value({"field":<null>,"field_type":"alias","rule_type":"matches","value":<null>})].field
│ attribute as the provider has marked it as required.
```

(!) For the historical data: Verified that KPI Id regenerated only in case KPI BS Id changed or KPI BS metric changed
(!) For the training data: Verified that threshold levels are mapped back

### CODE REVIEW

File: provider/provider_v6.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ import (
 
 const (
 	clientConcurrency = 15
-	defaultTimeout    = 60
+	defaultTimeout    = 220
```
Comment (benoittoulme) [inline provider/provider_v6.go:22]: why is that needed?

Comment (masha1997) [reply provider/provider_v6.go:22]: Sorry, forgot to remove the local change

File: provider/resource_service.go [hunk 2]
```diff
@@ -434,57 +722,26 @@ func service(ctx context.Context, d *schema.ResourceData, clientConfig models.Cl
...
-			}
-		}
-		err := d.Set("kpi", kpisNew)
-		if err != nil {
-			return nil, err
+	itsiKpis := []map[string]interface{}{}
+	for _, kpi := range m.KPIs {
+		if kpi.ID.IsUnknown() {
+			uuid, _ := uuid.GenerateUUID()
+			kpi.ID = types.StringValue(uuid)
```
Comment (benoittoulme) [inline provider/resource_service.go:729]: we can't use getKpiHashKey to get the KPI ID anymore?

Comment (masha1997) [reply provider/resource_service.go:729]: We can, we just didn't do this in prev version:
```
			if existingKpiId, ok := kpiOldKeys[internalIdentifier]; ok {
				kpiData["id"] = existingKpiId
			} else {
				kpiData["id"], _ = GenerateUUID(internalIdentifier)
			}
```

Comment (masha1997) [reply provider/resource_service.go:729]:
```
// testOverridable
var GenerateUUID = func(internalIdentifier string) (string, error) {
	return uuid.GenerateUUID()
}
```

Comment (masha1997) [reply provider/resource_service.go:729]: getKpiHashKey was used for internal mapping (it is still there in the plan modifier https://github.com/TiVo/terraform-provider-splunk-itsi/pull/418/files#diff-6bcac826cdcc1c2d664f422c46abbafc7b31fd82f90dfdb8ca063134fa250af2R343)

Comment (masha1997) [reply provider/resource_service.go:729]: AFAIK it was done that way, in order to follow "We don't know how it is saved and used in Mongo, if UI generates UUID, lets follow that strategy"

Comment (benoittoulme) [reply provider/resource_service.go:729]: got it. modifyplan does this instead, I forgot how the new plugin does this.

File: provider/resource_service.go [hunk 3]
```diff
@@ -144,262 +129,565 @@ func getKpiBSData(ctx context.Context, cc models.ClientConfig, id string) (map[s
...
+							Default:     stringdefault.StaticString("kpis_primary"),
+							Description: "Could be service_health or kpis_primary.",
+							Validators: []validator.String{
+								stringvalidator.OneOf("kpis_primary", "service_health"),
+							},
+						},
+						"urgency": schema.Int64Attribute{
+							Optional: true,
+							Computed: true,
+							//Default:     int64default.StaticInt64(5),
```
[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:194]: Thank you, I am very sorry I didn't leave a comment for that hook, it is for the import case, and the default urgency overrides the one specified in the config. (But it works ok, for the regular updates). I dig into that a bit, and looks like it for some reason doesn't see that int value by path in the config https://github.com/hashicorp/terraform-plugin-framework/blob/main/internal/fwschemadata/data_default.go#L83

But the generated file and all state structures from the import/read calls contain a proper value. Just commenting out default makes plan clean. Then I move the default logic, to the plan modifier.

File: provider/resource_service.go [hunk 4]
```diff
@@ -144,262 +129,565 @@ func getKpiBSData(ctx context.Context, cc models.ClientConfig, id string) (map[s
...
+	for _, kpi := range state.KPIs {
+
+		// kpiid is important to save for historical raw data. Historical raw data makes sense,
+		// until base search & metris is same
+		internalIdentifier := ""
+		resp.Diagnostics.Append(getKpiHashKey(kpi, &internalIdentifier)...)
+
+		kpiOldKeys[internalIdentifier] = &KpiMapFields{
+			ID:      kpi.ID,
+			Urgency: types.Int64Value(5),
```
[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:363]: Sorry, I didn't get that, why is it better? just to replace with const?

File: provider/resource_service.go [hunk 5]
```diff
@@ -1,123 +1,108 @@
...
+	Title                                 types.String             `json:"title" tfsdk:"title"`
+	Description                           types.String             `json:"description" tfsdk:"description"`
+	Enabled                               types.Bool               `json:"enabled" tfsdk:"enabled"`
+	IsHealthscoreCalculateByEntityEnabled types.Bool               `json:"is_healthscore_calculate_by_entity_enabled" tfsdk:"is_healthscore_calculate_by_entity_enabled"`
+	SecurityGroup                         types.String             `json:"sec_grp" tfsdk:"security_group"`
+	Tags                                  types.Set                `tfsdk:"tags"`
+	ShkpiID                               types.String             `json:"shkpi_id" tfsdk:"shkpi_id"`
+	KPIs                                  []*KpiState              `tfsdk:"kpi"`
+	EntityRules                           []*EntityRuleState       `tfsdk:"entity_rules"`
+	ServiceDependsOn                      []*ServiceDependsOnState `tfsdk:"service_depends_on"`
```
[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:64]: No special reason, I just browsed through examples of people working with blocks in the schema. A lot of used array of pointers, I decided it would be easier to work with them through the chain of functions calls, and that's all :)

File: provider/resource_service.go [hunk 6]
```diff
@@ -144,262 +129,565 @@ func getKpiBSData(ctx context.Context, cc models.ClientConfig, id string) (map[s
...
+				Description: "Set the Service Health Score calculation to account for the severity levels of individual entities if at least one KPI is split by entity.",
+			},
+			"security_group": schema.StringAttribute{
+				Optional:    true,
+				Description: "The team the object belongs to.",
+			},
+			"tags": schema.SetAttribute{
+				Optional:    true,
+				Description: "The tags for the service.",
+				ElementType: types.StringType,
```
[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:309]: https://github.com/TiVo/terraform-provider-splunk-itsi/pull/418/files#diff-6bcac826cdcc1c2d664f422c46abbafc7b31fd82f90dfdb8ca063134fa250af2R531

[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:309]: I'll test with defaults as well, I didn't realize it could be set for collections, just remembered it was not possible for sdkv2

File: provider/resource_service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:64]: It's not a huge deal, but I think it's usually better to avoid pointers unless they are actually necessary. Typical use case for pointers in my view is either 
1) when we want to pass an object by reference
2) if we want to support `null` values. 
This also makes me wonder, what would happen if a user puts a null value in one of those lists? Would terraform validate/reject it, or would it actually parse it as a null pointer, that we would then have to catch in the code? It's probably better not to have to worry about that, if possible :)

File: provider/resource_service.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:309]: Yes, it's possible in the new framework, for example: https://github.com/TiVo/terraform-provider-splunk-itsi/blob/v2-dev/provider/resource_entity.go#L242

Can do similar for maps, too.

[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:309]: But I was actually talking more generally about all the optional fields, not just tags. From what I remember it's usually better to set the default values even for simple types like Strings, etc. That should help eliminate drift when terraform config changes between `""` <-> `null`, for example..

[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:309]: But I also know that the new terraform framework has some bugs related to computed + optional fields, unfortunately. And sometimes we can't really use them without triggering those 🐛 s...

File: provider/resource_service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:194]: Got it, thanks for the  details. Yes, that definitely sounds like something worth mentioning in a comment 😄

File: provider/resource_service.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:363]: I presume `5` is defining the default urgency here? It's usually a good practice not to have magical numbers/strings in the code, especially if they are referenced more than once. I know it's only used once here, but I still think it's worth it to define it as a const just for code readability/maintainability purpose.

[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:363]: Basically the idea is to help code reader understand the significance of that number, so they don't have to guess its meaning. Not a big deal in this case, of course..

File: provider/resource_service.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:309]: Sorry, I re-reading in the morning, and realized, I inserted a bad link. I tried to point to:
https://github.com/TiVo/terraform-provider-splunk-itsi/blob/73c062c4f3836bbef3917ce01bcc52516c0ab051/provider/resource_service.go#L531
Thet in case len=0 I set NullSet, thank you, and I'll test with default :)

File: provider/resource_service.go [hunk 7]
```diff
@@ -163,34 +166,41 @@ func (r *resourceService) Schema(ctx context.Context, req resource.SchemaRequest
...
-							Optional:    true,
 							Description: "User-defined description for the KPI. ",
+							Optional:    true,
 						},
 						"type": schema.StringAttribute{
 							Optional:    true,
 							Computed:    true,
 							Default:     stringdefault.StaticString("kpis_primary"),
-							Description: "Could be service_health or kpis_primary.",
+							Description: "Could be kpis_primary.",
```
Comment (masha1997) [inline provider/resource_service.go:185]: We never support service_health

File: provider/resource_service.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:327]: Added defaults to everything, except empty string. It stands for the import issue also. If an empty string is specified it is mapped to null in configuration (even in case the passed state contains an empty string inside), and then it tries to compare null with "", and doesn't print that (nothing like description: null -> ""), just results in very unobvious plan

File: provider/resource_service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (masha1997) [reply provider/resource_service.go:67]: Is this ok I'll try to do this refactor in the scope of the task with threshold templates. Cause I should re-test carefully after

[the remark this answers is not in the document]
Comment (dmitrykruglov) [reply provider/resource_service.go:67]: Not sure what's the task regarding threshold templates, but yeah, if you wanna do that later in another PR, that's totally fine

**Review by dmitrykruglov — APPROVED** [review]
Looks good, thank you
